File: flowersec-ts/src/e2ee/transcript.ts
import { sha256 } from "@noble/hashes/sha256";
import { concatBytes, u16be, u32be } from "../utils/bin.js";

// TranscriptInputs contains the canonical fields hashed into the transcript.
export type TranscriptInputs = Readonly<{
  /** Protocol version byte used in the transcript. */
  version: number;
  /** Numeric suite identifier (see e2ee suite). */
  suite: number;
  /** Role byte (client=1, server=2). */
  role: number;
  /** Client feature bitset. */
  clientFeatures: number;
  /** Server feature bitset. */
  serverFeatures: number;
  /** Channel identifier shared by both endpoints. */
  channelId: string;
  /** Client nonce (32 bytes). */
  nonceC: Uint8Array; // 32
  /** Server nonce (32 bytes). */
  nonceS: Uint8Array; // 32
  /** Client ephemeral public key bytes. */
  clientEphPub: Uint8Array;
  /** Server ephemeral public key bytes. */
  serverEphPub: Uint8Array;
}>;

const te = new TextEncoder();

// transcriptHash computes the SHA-256 hash of the handshake transcript.
export function transcriptHash(inputs: TranscriptInputs): Uint8Array {
  if (inputs.nonceC.length !== 32 || inputs.nonceS.length !== 32) throw new Error("nonce must be 32 bytes");
  const channelIdBytes = te.encode(inputs.channelId);
  if (channelIdBytes.length > 0xffff) throw new Error("channel_id too long");
  if (inputs.clientEphPub.length > 0xffff || inputs.serverEphPub.length > 0xffff) throw new Error("pub too long");

  const prefix = te.encode("flowersec-e2ee-v1");
  const body = concatBytes([
    prefix,
    new Uint8Array([inputs.version & 0xff]),
    u16be(inputs.suite),
    new Uint8Array([inputs.role & 0xff]),
    u32be(inputs.clientFeatures >>> 0),
    u32be(inputs.serverFeatures >>> 0),
    u16be(channelIdBytes.length),
    channelIdBytes,
    inputs.nonceC,
    inputs.nonceS,
    u16be(inputs.clientEphPub.length),
    inputs.clientEphPub,
    u16be(inputs.serverEphPub.length),
    inputs.serverEphPub
  ]);
  return sha256(body);
}

File: flowersec-go/crypto/e2ee/transcript.go
package e2ee

import (
	"crypto/sha256"
	"errors"

	"github.com/floegence/flowersec/flowersec-go/internal/bin"
)

// ErrInvalidTranscriptInput signals a missing or oversized transcript field.
var ErrInvalidTranscriptInput = errors.New("invalid transcript input")

// TranscriptInputs captures the deterministic fields hashed into the transcript.
type TranscriptInputs struct {
	Version        uint8    // Protocol version byte.
	Suite          uint16   // Numeric suite identifier.
	Role           uint8    // Role byte (client/server).
	ClientFeatures uint32   // Client feature bitset.
	ServerFeatures uint32   // Server feature bitset.
	ChannelID      string   // Shared channel identifier.
	NonceC         [32]byte // Client nonce.
	NonceS         [32]byte // Server nonce.
	ClientEphPub   []byte   // Client ephemeral public key bytes.
	ServerEphPub   []byte   // Server ephemeral public key bytes.
}

// TranscriptHash computes the SHA-256 hash of the canonical handshake transcript.
func TranscriptHash(in TranscriptInputs) ([32]byte, error) {
	if in.ChannelID == "" {
		return [32]byte{}, ErrInvalidTranscriptInput
	}
	if len(in.ClientEphPub) == 0 || len(in.ServerEphPub) == 0 {
		return [32]byte{}, ErrInvalidTranscriptInput
	}

	channelIDBytes := []byte(in.ChannelID)
	if len(channelIDBytes) > 0xffff {
		return [32]byte{}, ErrInvalidTranscriptInput
	}
	if len(in.ClientEphPub) > 0xffff || len(in.ServerEphPub) > 0xffff {
		return [32]byte{}, ErrInvalidTranscriptInput
	}

	// transcript =
	//   "flowersec-e2ee-v1" ||
	//   version:u8 ||
	//   suite:u16be ||
	//   role:u8 ||
	//   client_features:u32be ||
	//   server_features:u32be ||
	//   channel_id_len:u16be || channel_id ||
	//   nonce_c(32) || nonce_s(32) ||
	//   client_pub_len:u16be || client_pub ||
	//   server_pub_len:u16be || server_pub
	prefix := []byte("flowersec-e2ee-v1")
	size := len(prefix) + 1 + 2 + 1 + 4 + 4 + 2 + len(channelIDBytes) + 32 + 32 + 2 + len(in.ClientEphPub) + 2 + len(in.ServerEphPub)
	buf := make([]byte, 0, size)
	buf = append(buf, prefix...)
	buf = append(buf, in.Version)
	tmp := make([]byte, 8)
	bin.PutU16BE(tmp[:2], in.Suite)
	buf = append(buf, tmp[:2]...)
	buf = append(buf, in.Role)
	bin.PutU32BE(tmp[:4], in.ClientFeatures)
	buf = append(buf, tmp[:4]...)
	bin.PutU32BE(tmp[:4], in.ServerFeatures)
	buf = append(buf, tmp[:4]...)
	bin.PutU16BE(tmp[:2], uint16(len(channelIDBytes)))
	buf = append(buf, tmp[:2]...)
	buf = append(buf, channelIDBytes...)
	buf = append(buf, in.NonceC[:]...)
	buf = append(buf, in.NonceS[:]...)
	bin.PutU16BE(tmp[:2], uint16(len(in.ClientEphPub)))
	buf = append(buf, tmp[:2]...)
	buf = append(buf, in.ClientEphPub...)
	bin.PutU16BE(tmp[:2], uint16(len(in.ServerEphPub)))
	buf = append(buf, tmp[:2]...)
	buf = append(buf, in.ServerEphPub...)

	return sha256.Sum256(buf), nil
}
